### ISSUE TITLE
SolrEndpoint - consolidate methods?

### ISSUE THREAD

**Issue by darrenleeweber** [projecthydra-labs/hyku#1071]
`CleanupAccountJob#cleanup_solr` might belong on `SolrEndpoint` (possibly a `before_destroy` callback).

**Comment by atz** [comment]
Except that solr is a remote service and therefore more than a simple ActiveRecord model should deal with.

**Comment by darrenleeweber** [comment]
Perhaps the advantage to managing connections to the endpoint in the SolrEndpoint is that the AR model will be within the tenant context; but that doesn't apply because the apartment initializer excludes them, i.e.:
```
config.excluded_models = %w{ Account Endpoint SolrEndpoint FcrepoEndpoint RedisEndpoint }
```
This seems to imply that all the Endpoint models have global (vs. tenant) app-scope.  So these models and anything that use them must do extra work to get the params for the tenant collections correct.

BTW, what seems to be a bit confusing (for my noob perspective), is whether the Endpoints all share the same service(s) (like a single Solr URL) and just manage collections in those services, or if they access tenant-sandboxes services (like each tenant is it's own VM/docker container with a suite of it's own services, which all have the same collection/schema within them, just different connection params).

There is also some confusing inconsistency, e.g. the `CleanupAccountJob` contains:
```
    # Solr endpoint pulls its connection info from account directly
    cleanup_solr(account)
    # Ensure Fedora endpoint's connection established to this account
    # and ensure Redis::Namespace limited to this account's namespace
    account.switch do
      cleanup_fedora(account)
      cleanup_redis(account)
    end
```
Here, solr cleanup is done outside the tenant swap, but everything else is inside the tenant swap.  Together with the apparent app-global scope of all the Endpoints, this is really confusing (for me at least).  Rhetorical question: what is special or different about the SolrEndpoint that it has privileged information about the tenant that the other Endpoint do not?

**Comment by cbeer** [comment]
> This seems to imply that all the Endpoint models have global (vs. tenant) app-scope. So these models and anything that use them must do extra work to get the params for the tenant collections correct.

Indeed, they are globally scoped. We did this believing they are a more-likely target for global management ("as an application administrator, I want to see the status of / update the schemas for / etc all the solr endpoints") compared to stuff like hyrax-managed data. The extra work is minimal (and handled by the apartment framework) and seems like it works well enough.

> There is also some confusing inconsistency, e.g. the `CleanupAccountJob`

I agree, that seems unfortunate. I suspect there are other ways to write those steps to avoid entering the tenant context or to improve the consistency of the code. I'm not sure that's a productive or valuable exercise right now, though, until there are specific value-adding changes that need to be made.

**Comment by darrenleeweber** [comment]
When `SolrEndpoint` has global app-context and it does a `.reset!` that appears to call class methods on `ActiveFedora::SolrService.reset!` - does that impact the entire HyKu app or just the tenant?  It seems like it would impact the entire app.  If I understand that correctly (probably not), I'm not sure what implications that has for application threading/scaling and access for a Site/Instance to it's collection data.  (Perhaps the `SolrService` that I'm thinking about already exists in `ActiveFedora::SolrService`, but that was never meant to deal with tenants and switching, so it gets wrapped up in the Endpoint to do that).

**Comment by cbeer** [comment]
`reset!` will only affect the current thread.
